### PR TITLE
feat/gene: add types for UTR5, CDS, UTR3

### PR DESCRIPTION
**Pull request by mnsmar**
A draft API following the discussion after https://github.com/biogo/biogo/pull/41#issuecomment-225650607.

### CODE REVIEW

File: feat/gene/gene.go [hunk 1]
```diff
@@ -285,6 +314,38 @@ func (t *CodingTranscript) SetExons(exons ...Exon) error {
 	return nil
 }
 
+// TranscriptFeature defines a feature on a transcript.
+type TranscriptFeature struct {
+	Transcript Transcript
```
Comment (kortschak) [inline feat/gene/gene.go]: Can you add doc comments to the fields here please.

File: feat/gene/gene.go [hunk 2]
```diff
@@ -285,6 +314,38 @@ func (t *CodingTranscript) SetExons(exons ...Exon) error {
...
+// Start returns the feature start relative to Transcript.
+func (t *TranscriptFeature) Start() int { return t.Offset }
+
+// End returns the feature end relative to TranscriptLocation.
+func (t *TranscriptFeature) End() int { return t.Offset + t.Length }
+
+// Len returns the length of the feature.
+func (t *TranscriptFeature) Len() int { return t.Length }
+
+// Name returns an empty string.
```
Comment (kortschak) [inline feat/gene/gene.go:330]: Why?

Comment (mnsmar) [reply feat/gene/gene.go:330]: I couldn't think of a good reason to add an extra field for name so I followed what we did for Exon below. I will add it now.

**Comment by mnsmar** [comment]
Note that I marked the `UTR5Start`, etc as deprecated.

**Comment by mnsmar** [comment]
Any other comments on this? If the API looks good I will go on and refactor the tests to test new functionality.

File: feat/gene/gene.go [hunk 3]
```diff
@@ -218,52 +218,85 @@ func (t *CodingTranscript) Orientation() feat.Orientation {
...
 	case feat.Forward:
-		return 0
+		start = 0
+		end = t.CDSstart
 	case feat.Reverse:
-		return t.CDSend
+		start = t.CDSend
+		end = t.Len()
+	default:
+		panic("gene: zero orientation for transcript")
```
Comment (kortschak) [inline feat/gene/gene.go]: s/zero/invalid/

File: feat/gene/gene.go [hunk 4]
```diff
@@ -218,52 +218,85 @@ func (t *CodingTranscript) Orientation() feat.Orientation {
...
 	case feat.Forward:
-		return t.CDSstart
+		start = t.CDSend
+		end = t.Len()
 	case feat.Reverse:
-		return t.Len()
+		start = 0
+		end = t.CDSstart
+	default:
+		panic("gene: zero orientation for transcript")
```
Comment (kortschak) [inline feat/gene/gene.go]: s/zero/invalid/

File: feat/gene/gene.go [hunk 5]
```diff
@@ -285,6 +318,39 @@ func (t *CodingTranscript) SetExons(exons ...Exon) error {
 	return nil
 }
 
+// TranscriptFeature defines a feature on a transcript.
+type TranscriptFeature struct {
+	Transcript Transcript       // Transcript is the transcript that the feature is located.
+	Offset     int              // Offset is the position of the feature relative to Transcript.
```
Comment (kortschak) [inline feat/gene/gene.go:314]: Is there a reason to keep {offset, length} rather than {start, end}?

File: feat/gene/gene.go [hunk 6]
```diff
@@ -218,52 +218,85 @@ func (t *CodingTranscript) Orientation() feat.Orientation {
...
+		Transcript: t,
+		Offset:     start,
+		Length:     end - start,
+		Orient:     feat.Forward,
 	}
-	panic("gene: zero orientation for transcript")
+}
+
+// UTR5start returns the start of the 5'UTR relative to the transcript.
+// Deprecated: New code should use t.UTR5().Start() instead.
```
Comment (kortschak) [inline feat/gene/gene.go]: Deprecation indicates that these methods will be removed at a later date. This won't happen, so please rewrite this as something like "UTR5start is shorthand for UTR5().Start()." Same for the methods below.

File: feat/gene/gene.go (hunk 5 above)
Comment (mnsmar) [reply feat/gene/gene.go:314]: Sorry for the late reply. Just for consistency with the other types. Should I change it?

Comment (kortschak) [reply feat/gene/gene.go:314]: No, it's fine like that in the context of the other types.

**Comment by mnsmar** [comment]
Addressed the comments and added tests.

**Review by kortschak — APPROVED** [review]
Thank you.

**Pull request merged**